### PR TITLE
Ports: Add dmidecode

### PR DESCRIPTION
**Pull request by supercomputer7**
For this to work, I added two entries in `ProcFS`, and also added a new device called `MemoryDevice` (`/dev/mem`), which only allows currently to map the BIOS ROM between physical address `0xf0000` to `0x100000`.

### CODE REVIEW

File: Ports/AvailablePorts.md [hunk 1]
```diff
@@ -69,3 +69,4 @@ Please make sure to keep this list up to date when adding and updating ports. :^
 | [`vttest`](vttest/)            | vttest                                        | 20200610          | https://invisible-island.net/vttest/                  |
 | [`zlib`](zlib/)                | zlib                                          | 1.2.11            | https://www.zlib.net/                                 |
 | [`zstd`](zstd/)                | Zstandard                                     | 1.4.4             | https://facebook.github.io/zstd/                      |
+| [`dmidecode`](dmidecode/)                | dmidecode                                     | 3.1             | https://github.com/mirror/dmidecode                      |
```
Comment (linusg) [inline Ports/AvailablePorts.md:72]: Please keep this list sorted ;)

Comment (supercomputer7) [reply Ports/AvailablePorts.md]: Thanks for noticing :)

**Comment by supercomputer7** [comment]
I made `dmidecode` to build with less patches than before, essentially now the patch files are semantic changes :)

**Comment by supercomputer7** [comment]
It seems like `biosdecode` can use `/dev/mem` just fine, but `dmidecode` fails when using `/dev/mem`.

**Comment by supercomputer7** [comment]
After couple of hours of debugging I finally figured it was a declaration of enum with wrong order of values... the `sysconf` syscall returned the wrong value, therefore a `mmap` called by `dmidecode` failed, what a journey!

**Comment by supercomputer7** [comment]
I've added a better mechanism to determine if userspace is allowed to `mmap` physical memory with `/dev/mem` - in boot we will add all multiboot physical ranges to a Vector, then when trying to mmap, we will check if the `PhysicalAddress` is within a known range, and if so, allow to `mmap` for reserved ranges only. If we didn't find a range that corresponds to the checked address, we also allow to `mmap` it.

Also, this PR is related to https://github.com/SerenityOS/serenity/issues/2599, as I suggested to remove the older `DMIDecoder` component and expose the DMI blobs in `/proc/`.

File: Kernel/Devices/MemoryDevice.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2018-2020, Andreas Kling <kling@serenityos.org>
```
Comment (awesomekling) [inline Kernel/Devices/MemoryDevice.cpp:2]: I didn't write this :)

Comment (supercomputer7) [reply Kernel/Devices/MemoryDevice.cpp]: Oops, my bad!

**Review by awesomekling — CHANGES_REQUESTED** [review]

File: Kernel/UnixTypes.h [hunk 3]
```diff
@@ -60,8 +60,8 @@ enum {
     _SC_NPROCESSORS_CONF,
     _SC_NPROCESSORS_ONLN,
     _SC_OPEN_MAX,
-    _SC_PAGESIZE,
     _SC_TTY_NAME_MAX,
+    _SC_PAGESIZE,
```
[the remark this answers is not in the document]
Comment (supercomputer7) [reply Kernel/UnixTypes.h:64]: At least it works correctly now, `dmidecode` can `mmap` happily and I'm happy too :)

File: Kernel/DMI.h [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+#include <AK/Types.h>
+#include <AK/Vector.h>
+#include <Kernel/KBuffer.h>
+#include <Kernel/PhysicalAddress.h>
+#include <Kernel/VM/Region.h>
+#include <Kernel/VirtualAddress.h>
+
+namespace Kernel {
+
+namespace SMBIOS {
```
[the remark this answers is not in the document]
Comment (supercomputer7) [reply Kernel/DMI.h:39]: `DMIExpose` class is outside of this namespace on intent. But yeah, one empty line after opening a namespace is a good thing :)

File: Kernel/Devices/MemoryDevice.cpp [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+{
+}
+
+MemoryDevice::~MemoryDevice()
+{
+}
+
+bool MemoryDevice::is_allowed_range(PhysicalAddress offset, const Range& range) const
+{
+    return MM.is_allowed_to_mmap_to_userspace(PhysicalAddress(offset.get())) && MM.is_allowed_to_mmap_to_userspace(PhysicalAddress(offset.get() + PAGE_ROUND_UP(range.size())));
```
[the remark this answers is not in the document]
Comment (supercomputer7) [reply Kernel/Devices/MemoryDevice.cpp:47]: can we `mmap` with a range of not page size multiples? from what I understand we can't and therefore I round it with the macro.

[the remark this answers is not in the document]
Comment (awesomekling) [reply Kernel/Devices/MemoryDevice.cpp:47]: The rounding is fine, but (base+size) is not the last byte of the range, it's one byte *after* the range.

[the remark this answers is not in the document]
Comment (supercomputer7) [reply Kernel/Devices/MemoryDevice.cpp]: Oh I see, I was wrong about this, indeed it should take base+size so we don't have that "1 byte after" problem.

File: Ports/AvailablePorts.md [hunk 6]
```diff
@@ -12,6 +12,7 @@ Please make sure to keep this list up to date when adding and updating ports. :^
 | [`curl`](curl/)                | curl                                          | 7.65.3            | https://curl.se/                                      |
 | [`dash`](dash/)                | DASH                                          | 0.5.10.2          | http://gondor.apana.org.au/~herbert/dash              |
 | [`diffutils`](diffutils/)      | GNU Diffutils                                 | 3.5               | https://www.gnu.org/software/diffutils/               |
+| [`dmidecode`](dmidecode/)      | dmidecode                                     | 3.3             | https://github.com/mirror/dmidecode                      |
```
Comment (bcoles) [inline Ports/AvailablePorts.md:15]: Fix table formatting alignment for consistency:

```suggestion
| [`dmidecode`](dmidecode/)      | dmidecode                                     | 3.3               | https://github.com/mirror/dmidecode                   |
```

File: Ports/dmidecode/package.sh [hunk 7]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env -S bash ../.port_include.sh
+port=dmidecode
+version=3.3
+useconfigure=false
+prefix=
+files="http://download.savannah.gnu.org/releases/dmidecode/dmidecode-${version}.tar.xz dmidecode-${version}.tar.xz"
```
Comment (bcoles) [inline Ports/dmidecode/package.sh:6]: The archive file should be validated using the provided signature. The signature file (preferably both the archive and signature) should also be downloaded over HTTPS.

Untested, but this should do the trick:

```suggestion
files="https://download.savannah.gnu.org/releases/dmidecode/dmidecode-${version}.tar.xz dmidecode-${version}.tar.xz
https://download.savannah.gnu.org/releases/dmidecode/dmidecode-${version}.tar.xz.sig dmidecode-${version}.tar.xz.sig"
auth_type="sig"
auth_opts="--keyring ./gnu-keyring.gpg dmidecode-${version}.tar.xz.sig"
```

File: Ports/dmidecode/patches/dmiopt.h.patch [hunk 8]
```diff
@@ -0,0 +1,13 @@
+diff --git a/dmiopt.h b/dmiopt.h
+index 2374637..d215aff 100644
+--- a/dmiopt.h
++++ b/dmiopt.h
+@@ -45,7 +45,7 @@ extern struct opt opt;
+ #define FLAG_QUIET              (1 << 3)
+ #define FLAG_DUMP_BIN           (1 << 4)
+ #define FLAG_FROM_DUMP          (1 << 5)
+-#define FLAG_NO_SYSFS           (1 << 6)
++#define FLAG_NO_PROCFS           (1 << 6)
```
Comment (bcoles) [inline Ports/dmidecode/patches/dmiopt.h.patch:10]:
```suggestion
+#define FLAG_NO_PROCFS          (1 << 6)
```

File: Ports/AvailablePorts.md (hunk 6 above)
Comment (supercomputer7) [reply Ports/AvailablePorts.md:15]: Thanks for noticing! just asking - is there a tool to align these columns correctly so I can do that next time?

Comment (bcoles) [reply Ports/AvailablePorts.md:15]: None that I'm aware of.

**Pull request merged**